src/solvers/memory_solution_history.C

```diff
@@ -128,10 +128,47 @@ void MemorySolutionHistory::store(bool /* is_adjoint_solve */)
     saved_vectors[_solution] = _system.solution->clone();
 }
 
-void MemorySolutionHistory::retrieve(bool /* is_adjoint_solve */)
+void MemorySolutionHistory::retrieve(bool is_adjoint_solve)
 {
   this->find_stored_entry();
 
+  // To set the deltat while using adaptive timestepping, we will utilize
+  // consecutive time entries in the stored solutions iterator
+  Real _current_time = stored_sols->first;
+
+  // If we are solving the adjoint, we are moving backwards, so decrement time
+  // else we are moving forwards, so increment time
+  if(is_adjoint_solve)
+  {
+    stored_solutions_iterator stored_sols_decrement_time = stored_sols;
+
+    // Recovering deltats needs two different entries from the the
+    // stored solutions map
+    if(stored_sols_decrement_time != stored_solutions.begin())
+    {
+      stored_sols_decrement_time--;
+
+      Real _decremented_time = stored_sols_decrement_time->first;
+
+      dynamic_cast<DifferentiableSystem &>(_system).deltat = _current_time - _decremented_time;
```

Comment:
This breaks MemorySolutionHistory compatibility with any non-diffsystem.  There's no way to actually make adaptive adjoint time stepping work with those right now, but we should at least be catching std::bad_cast (or using a dynamic_cast of pointer and checking for nullptr) and reverting to a uniform-timestepping code path to keep that working as generally as possible.

Reply:
Btw, is system the right owner for deltat, shouldnt the timesolver own it ?

Reply:
Hmm... IIRC I wanted the physics to be able to integrate without querying the time solver.  But yeah, this was a mistake.  I'd change it right now if I could figure out a nice temporarily-backwards-compatibility-preserving path to do so.